### ISSUE TITLE
Mapping atoms from constructed molecule to building blocks

### ISSUE THREAD

**Issue by joshkamm** [lukasturcani/stk#260]
Is there any available mechanism for mapping an atom in a constructed molecule to the corresponding atom in the building block that it was constructed from? If not, any pointers on how this could be done?

My initial use case is for linear polymers which may make things simpler but I could envision potentially going beyond that in the future.

If it would be helpful I can provide more details about what I'm trying to do.

P.S. I've been working on machine learning of good conformers of polymers and your library has been helpful. Thanks!

**Comment by lukasturcani** [comment]
Hi & thank you! Yes this is possible. 

```python

constructed = stk.ConstructedMolecule(...)
for atom_info in constructed.get_atom_infos():
    atom = atom_info.get_atom()
    # If the atom was added during construction, these will return None.
    building_block = atom_info.get_building()
    building_block_id = atom_info.get_building_block_id()

```

https://stk.readthedocs.io/en/latest/stk.molecular.molecules.constructed_molecule.html#stk.molecular.molecules.constructed_molecule.ConstructedMolecule.get_atom_infos

https://stk.readthedocs.io/en/latest/stk.molecular.atoms.atom_info.html#stk.molecular.atoms.atom_info.AtomInfo

Key point is between `building_block` and `building_block_id`. Since each building block can be placed multiple times during the construction just returning the original building block might not be all the info you want. So building block id returned a unique id for every placement of a building block. For polymers, this means that you can identify each monomer, rather than just all the atoms in the constructed molecule which came from the original building block, which could be half of all the atoms in the polymer. (I feel like I always struggle to explain this, so let me know if its not clear).

There is also `BondInfo`, which gives you the same data but for bonds.

https://stk.readthedocs.io/en/latest/stk.molecular.molecules.constructed_molecule.html#stk.molecular.molecules.constructed_molecule.ConstructedMolecule.get_bond_infos

https://stk.readthedocs.io/en/latest/stk.molecular.bonds.bond_info.html#stk.molecular.bonds.bond_info.BondInfo

Might be good to include this example in the basic examples actually.

**Comment by joshkamm** [comment]
Thanks for the informative response!

I think the building_block vs building_block_id makes sense partially because I already have to think about that distinction. I'm trying to use the atoms to map torsions between a polymer and its building blocks. An individual torsion could include atoms that span multiple occurrences of the same building block.

I see that the id of the atom returned seems like the index of the atom in the constructed molecule. Is there a way to tell what the corresponding atom id was in the building block (or equivalently, which atom within the building block)?

**Comment by lukasturcani** [comment]
Hmm, I don't think this is easy to get atm, but I think it should be easy to add. I'll put it on the to-do list.

**Comment by joshkamm** [comment]
Okay thanks!

In the meantime, it seems like for linear polymers the atom indices follow my intuition of being in contiguous chunks corresponding to each building block in order along the polymer, and that within a building block the ordering of atom indices is maintained. If you know of cases when that wouldn't be the case let me know, but otherwise that should be enough for me to get started.

**Comment by lukasturcani** [comment]
Yep, the order of ids of atoms  in the constructed molecule should always follow the order building block (for a set of atoms which originate from the same building block placement). The main thing to look out for is that some atoms in the building block may be removed in the constructed molecule, so while the ordering of ids is maintained, the difference between ids may not be. So for example, atom 1 and atom 5 in the building block, could be atom 1 and atom 3 in the constructed molecule, if atom 2 and atom 3 in the original building block got removed during construction.

**Comment by lukasturcani** [comment]
Should be solved by #267.

I've also made a new release of `stk`, so assuming you installed it with `pip`, you can run `pip install stk --upgrade` to get the latest version. Release notes are here https://github.com/lukasturcani/stk/releases but the most important bit is that you can now use `AtomInfo.get_building_block_atom`, which will return the atom the building block, for a given constructed molecule atom.

**Comment by joshkamm** [comment]
Wow thanks! This is the first feedback I've given on an open source project outside my research group that I'm aware was acted on.

I did install with pip. I ran the `pip install stk --upgrade` and was able to access `AtomInfo.get_building_block_atom()`. I can let you know if I run into issues but I experimented on a small polymer that I'm currently using for testing, and the new method seems to behave as expected.
